### PR TITLE
Symex taint analysis

### PR DESCRIPTION
**Pull request by DanielNeville**
Feedback and corrections welcomed.

### CODE REVIEW

File: src/path-symex/path_symex_history.h [hunk 1]
```diff
@@ -9,8 +9,8 @@ Author: Daniel Kroening, kroening@kroening.com
 #ifndef CPROVER_PATH_SYMEX_HISTORY_H
 #define CPROVER_PATH_SYMEX_HISTORY_H
 
-#include <cassert>
 #include <limits>
+#include <cassert>
```
Comment (tautschnig) [inline src/path-symex/path_symex_history.h]: What's the rationale for this change?

File: src/goto-programs/remove_ternary.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+\*******************************************************************/
+
+#ifndef CPROVER_REMOVE_TERNARY_H
+#define CPROVER_REMOVE_TERNARY_H
+
+#include <goto-programs/goto_model.h>
+
+class remove_ternaryt
+{
```
Comment (tautschnig) [inline src/goto-programs/remove_ternary.h]: This class definition should probably be moved to the .cpp file as is done in other transformations.

File: src/goto-programs/remove_ternary.cpp [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+  /* LHS handles ternary when used on the LHS of an assignment
+   * operator.
+   * LHS = false is used whenever the ternary is simply
+   * part of an arbitrary condition (including all RHS)
+   */
+
+  if(expr.id() == ID_code)
+  {
+    codet &code=to_code(expr);
+    if(to_code(expr).get_statement() == ID_assign)
```
Comment (tautschnig) [inline src/goto-programs/remove_ternary.cpp]: Should use code instead of re-doing to_code(...)

File: src/goto-programs/remove_ternary.cpp [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+ Outputs:
+
+ Purpose:
+
+ \*******************************************************************/
+
+void remove_ternaryt::operator()(goto_functionst &goto_functions)
+{
+  Forall_goto_functions(f_it, goto_functions){
+  goto_programt &goto_program=f_it->second.body;
```
Comment (tautschnig) [inline src/goto-programs/remove_ternary.cpp]: Indentation?

File: src/goto-programs/remove_ternary.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+  void replace_ternary(
+      goto_programt &goto_program,
+      goto_programt::instructionst::iterator &i_it,
+          exprt &expr,
+          bool lhs);
+
+  void contains_ternary(exprt &expr, bool &contains);
+
+  bool contains_ternary(exprt &expr);
```
Comment (tautschnig) [inline src/goto-programs/remove_ternary.h]: Neither the "contains_ternary" nor "replaced" are used anywhere; the former is really dangerous as it will kill all sharing (well, actually contains_ternary would have to take a const exprt and be const).

File: src/path-symex/path_symex.cpp [hunk 6]
```diff
@@ -1055,6 +1084,144 @@ void path_symext::operator()(
...
+  if(!code.has_operands())
+    return;
+
+  assert(code.has_operands());
+
+  // Check if code is a set_taint CPROVER special function
+  if(code.op0().id() == ID_set_taint)
+  {
+    exprt src=code.op0();
+    assert(src.has_operands());
```
Comment (tautschnig) [inline src/path-symex/path_symex.cpp]: That seems insufficient: you actually want src.operands().size()>1

File: src/path-symex/path_symex.cpp [hunk 7]
```diff
@@ -1055,6 +1084,144 @@ void path_symext::operator()(
...
+
+    // Retrieve Symbol name.
+    assert(src.op0().id() == ID_address_of);
+    assert(src.op0().has_operands());
+    assert(src.op0().op0().id() == ID_index);
+    assert(src.op0().op0().has_operands());
+    assert(src.op0().op0().op0().id() == ID_string_constant);
+
+    dstring symbol_name=src.op0().op0().op0().get_string(ID_value);
```
Comment (tautschnig) [inline src/path-symex/path_symex.cpp]: Don't use dstring directly, use irep_idt

File: src/path-symex/path_symex_state.h [hunk 8]
```diff
@@ -13,15 +13,18 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "var_map.h"
 #include "path_symex_history.h"
 
+#include <path-symex/path_symex_taint_data.h>
```
Comment (tautschnig) [inline src/path-symex/path_symex_state.h]: Use "path_symex_taint_data.h" for consistency

File: src/path-symex/path_symex_state_read.cpp [hunk 9]
```diff
@@ -347,6 +347,66 @@ exprt path_symex_statet::instantiate_rec(
...
+     *  Fetch symbol name.
+     *
+     *  Note that string constants are rewritten into:
+     *    address_of -> index -> string_constant.
+     */
+    assert(src.op0().id() == ID_address_of);
+    assert(src.op0().op0().id() == ID_index);
+    assert(src.op0().op0().op0().id() == ID_string_constant);
+
+    dstring symbol_name=src.op0().op0().op0().get_string(ID_value);
```
Comment (tautschnig) [inline src/path-symex/path_symex_state_read.cpp]: dstring...

File: src/path-symex/path_symex_taint_data.cpp [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+ Module: Taint Data Module
+
+ Author: Daniel Neville, daniel.neville@cs.ox.ac.uk
+ John Galea,		john.galea@cs.ox.ac.uk
+
+ \*******************************************************************/
+
+#include "path_symex_taint_data.h"
+
+taint_datat::taint_rulet::taint_rulet()
```
Comment (tautschnig) [inline src/path-symex/path_symex_taint_data.cpp]: Use :loc(0), taint(0), ...

File: src/symex/path_search.cpp [hunk 11]
```diff
@@ -36,10 +36,16 @@ path_searcht::resultt path_searcht::operator()(
   
   locs.build(goto_functions);
 
-  // this is the container for the history-forest  
+  // Perform a non-comprehensive check on JSON file.
+  if(taint_engine->enabled &&
```
Comment (tautschnig) [inline src/symex/path_search.cpp]: Isn't this lacking a test for taint_engine being non-null?

File: src/symex/symex_parse_options.h [hunk 12]
```diff
@@ -41,6 +41,8 @@ class optionst;
   "(round-to-nearest)(round-to-plus-inf)(round-to-minus-inf)(round-to-zero)" \
   "(show-locs)(show-vcc)(show-properties)(show-goto-functions)" \
   "(property):(trace)(show-trace)(stop-on-fail)(eager-infeasibility)" \
+  "(taint):(show-taint-data)(taint-file):" \
+  "(remove-ternary)" \
```
Comment (tautschnig) [inline src/symex/symex_parse_options.h]: --help output for all of these, please

**Comment by DanielNeville** [comment]
Thanks Michael.  Appreciated.  Will sort.

File: src/symex/path_search.cpp (hunk 11 above)
Comment (DanielNeville) [inline src/symex/path_search.cpp]: Indeed.  Is there a preferred syntax within CProver for checking for null-ness?

(C++ supports a few methods)

**Comment by forejtv** [comment]
@DanielNeville can you rebase, please?

**Comment by hannes-steffenhagen-diffblue** [comment]
This is obsolete. Part of this apparently made it into the dependency analysis in VSD according to @chrisr-diffblue

**Comment by martin-cs** [comment]
At the moment VSD doesn't do taint analysis.  I think the larger issue is that the single-path symex was merged into the main CBMC.

**Comment by peterschrammel** [comment]
This seems to be for https://github.com/diffblue/symex, which was moved out of CBMC.
